speakeasy/binemu.py

```diff
@@ -804,59 +805,42 @@ def get_module_from_addr(self, addr):
                 return mod[0]
         return None
 
-    def get_api_hook(self, mod_name, func_name):
+    def get_api_hooks(self, mod_name, func_name) -> List[common.ApiHook]:
         """
         If an API hook has been set, return it here
         """
+        user_hooks = self.hooks.get(common.HOOK_API)
+        if not user_hooks:
+            return []
 
-        hooks = self.hooks.get(common.HOOK_API)
-        if not hooks:
-            return None
-
-        # See if we can quickly resolve the api hook
-        quick_look, wild_list = hooks
         api = (mod_name + '.' + func_name).lower()
-        qh = quick_look.get(api)
-        if qh:
-            return qh
-
-        # See if a wild card api hook was registered
-        for hook in wild_list:
-            if fnmatch.fnmatch(mod_name.lower(), hook.module.lower()):
-                if fnmatch.fnmatch(func_name.lower(), hook.api_name.lower()):
-                    return hook
-        return None
+        # The list always start with the most accurate
+        try:
+            found_hooks = user_hooks[api]
+        except KeyError:
+            found_hooks = []
+
+        # Retrieve every hook that was saved with a key that matches the request
+        for hook_api, hooks in user_hooks.items():
```

Comment:
Is there a way we can forgo entering this loop if there are no wild card lookups registered? The way it was implemented previously avoided this if we knew no wild cards were being used. This provides a significant speed boost when there are an extremely large amount of APIs that are hooked.

Reply:
A naive solution could be to add back the `wildcard` parameter in the `add_api_hook` method and, when a wild card is registered, we make the for loop check. If no wildcard are registered, we can stop the user hook search at line 821, allowing us to skip the slow for loop.

The solution that I implemented is a little bit complex, I don't really know if is the best solution, but is what came to my mind.

The basic idea is to try to identify _where_ the wildcard has been put.
The hooks will have the following struct if no wildcard has been set:
```json
hooks = ({ "kernel32" :  ( {"GetSystemTimeAsFileTime": [userhook1, userhook2]}, false), false)
```
With a wildcard at api level:
```json
hooks = ({ "kernel32" :  ( {"Virtual*": [userhook1, userhook2]}, true), false)
```
With a wildcard at module and api level:

```json
hooks = ({ "*" :  ( {"*": [userhook1, userhook2]}, true), true)
```

With this structure, we can optimize _how_ we search for the hooks.
- No module wildcard **AND**  No api wildcard ->  return hooks[module][api] 
- No module wildcard **BUT** api wildcard -> return [hook for api in hooks[module] if regex_ok(api, api_asked) for hook in api ]
I hope you get the point of what will happen if the module wildcard is enabled.